File: Arduino/epd2in13g/EPD_2in13g.cpp
/*****************************************************************************
* | File      	:   EPD_2in13g.c
* | Author      :   Waveshare team
* | Function    :   2inch13 e-paper (G)
* | Info        :
*----------------
* |	This version:   V1.0
* | Date        :   2023-05-29
* | Info        :
* -----------------------------------------------------------------------------
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documnetation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to  whom the Software is
# furished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS OR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
#
******************************************************************************/
#include "EPD_2in13g.h"
#include <stdlib.h>
Epd::~Epd() {
};
Epd::Epd() {
    reset_pin = RST_PIN;
    dc_pin = DC_PIN;
    cs_pin = CS_PIN;
    busy_pin = BUSY_PIN;
    WIDTH = EPD_2IN13G_WIDTH;
    HEIGHT = EPD_2IN13G_HEIGHT;
};
/******************************************************************************
function :	Software reset
parameter:
******************************************************************************/
void Epd::Reset(void)
{
    DigitalWrite(RST_PIN, 1);
    DelayMs(200);
    DigitalWrite(RST_PIN, 0);
    DelayMs(2);
    DigitalWrite(RST_PIN, 1);
    DelayMs(200);
}

/******************************************************************************
function :	send command
parameter:
     Reg : Command register
******************************************************************************/
void Epd::SendCommand(UBYTE Reg)
{
    DigitalWrite(DC_PIN, 0);
    DigitalWrite(CS_PIN, 0);   
    SpiTransfer(Reg);
    DigitalWrite(CS_PIN, 1);
}

/******************************************************************************
function :	send data
parameter:
    Data : Write data
******************************************************************************/
void Epd::SendData(UBYTE Data)
{
    DigitalWrite(DC_PIN, 1);
    DigitalWrite(CS_PIN, 0);
    SpiTransfer(Data);
    DigitalWrite(CS_PIN, 1);
}

/******************************************************************************
function :	Wait until the busy_pin goes HIGH
parameter:
******************************************************************************/
void Epd::ReadBusy(void)
{
    DelayMs(100);
    Serial.println("busy");
    while(DigitalRead(BUSY_PIN) != 1) {      //HIGH: idle, LOW: busy
        DelayMs(100);
    }
    Serial.println("idie");
}

/******************************************************************************
function :	Turn On Display
parameter:
******************************************************************************/
void Epd::TurnOnDisplay(void)
{

    SendCommand(0x12); // DISPLAY_REFRESH
    SendData(0x00);
    ReadBusy();

}

/******************************************************************************
function :	Setting the display window
parameter:
******************************************************************************/
void Epd::SetWindows(void)
{
    SendCommand(0x61); //TRES
    SendData(Source_BITS/256);		// EPD_2IN13G_WIDTH_H
    SendData(Source_BITS%256);		// EPD_2IN13G_WIDTH_L
    SendData(EPD_2IN13G_HEIGHT/256);			// EPD_2IN13G_HEIGHT_H
    SendData(EPD_2IN13G_HEIGHT%256); 		// EPD_2IN13G_HEIGHT_L	
}

/******************************************************************************
function :	Initialize the e-Paper register
parameter:
******************************************************************************/
int Epd::Init(void)
{
    if (IfInit() != 0) {
        return -1;
    }
    Reset();
    ReadBusy();
    SendCommand(0x4D);
    SendData(0x78);

    SendCommand(0x00);	//PSR
    SendData(0x0F);
    SendData(0x29);

    SendCommand(0x01);	//PWRR
    SendData(0x07);
    SendData(0x00);

    SendCommand(0x03);	//POFS
    SendData(0x10);
    SendData(0x54);
    SendData(0x44);

    SendCommand(0x06);	//BTST_P
    SendData(0x05);
    SendData(0x00);
    SendData(0x3F);
    SendData(0x0A);
    SendData(0x25);
    SendData(0x12);
    SendData(0x1A); 

    SendCommand(0x50);	//CDI
    SendData(0x37);

    SendCommand(0x60);	//TCON
    SendData(0x02);
    SendData(0x02);

    SetWindows();  // Setting the display window

    SendCommand(0xE7);
    SendData(0x1C);

    SendCommand(0xE3);	
    SendData(0x22);

    SendCommand(0xB4);
    SendData(0xD0);
    SendCommand(0xB5);
    SendData(0x03);

    SendCommand(0xE9);
    SendData(0x01); 

    SendCommand(0x30);
    SendData(0x08);  

    SendCommand(0x04);
    ReadBusy(); 
    return 0;

}

/******************************************************************************
function :	Clear screen
parameter:
******************************************************************************/
void Epd::Clear(UBYTE color)
{
    UWORD Width, Height;
    Width = Source_BITS / 4 ;
    Height = EPD_2IN13G_HEIGHT;

    SendCommand(0x10);
    for (UWORD j = 0; j < Height; j++) {
        for (UWORD i = 0; i < Width; i++) {
            SendData((color << 6) | (color << 4) | (color << 2) | color);
        }
    }

    TurnOnDisplay();
}

/******************************************************************************
function :	Sends the image buffer in RAM to e-Paper and displays
parameter:
******************************************************************************/
void Epd::Display(const UBYTE *Image)
{
    UWORD Width, Height;
    Width = (EPD_2IN13G_WIDTH % 4 == 0)? (EPD_2IN13G_WIDTH / 4 ): (EPD_2IN13G_WIDTH / 4 + 1);
    Height = EPD_2IN13G_HEIGHT;

    SendCommand(0x10);
    for (UWORD j = 0; j < Height; j++) {
        for (UWORD i = 0; i < (Source_BITS/4); i++) {
            if(i < 31 ) 
            SendData(pgm_read_byte(&Image[i + j * Width]));
            else SendData(0x00);
        }
    }
    TurnOnDisplay();
}

/******************************************************************************
function :	Enter sleep mode
parameter:
******************************************************************************/
void Epd::Sleep(void)
{
    SendCommand(0x02); //power off
    ReadBusy();       //waiting for the electronic paper IC to release the idle signal
    DelayMs(100);           //!!!The delay here is necessary,100mS at least!!! 
    
    SendCommand(0x07);  //deep sleep
    SendData(0XA5);
}

File: RaspberryPi_JetsonNano/python/lib/waveshare_epd/epd2in13g.py
# *****************************************************************************
# * | File        :	  epd2in13g.py
# * | Author      :   Waveshare team
# * | Function    :   Electronic paper driver
# * | Info        :
# *----------------
# * | This version:   V1.0
# * | Date        :   2023-05-29
# # | Info        :   python demo
# -----------------------------------------------------------------------------
# ******************************************************************************/
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documnetation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to  whom the Software is
# furished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS OR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
#

import logging
from . import epdconfig

import PIL
from PIL import Image
import io

# Display resolution
EPD_WIDTH       = 122
EPD_HEIGHT      = 250

logger = logging.getLogger(__name__)

class EPD:
    def __init__(self):
        self.reset_pin = epdconfig.RST_PIN
        self.dc_pin = epdconfig.DC_PIN
        self.busy_pin = epdconfig.BUSY_PIN
        self.cs_pin = epdconfig.CS_PIN
        self.width = EPD_WIDTH
        self.height = EPD_HEIGHT
        self.BLACK  = 0x000000   #   00  BGR
        self.WHITE  = 0xffffff   #   01
        self.YELLOW = 0x00ffff   #   10
        self.RED    = 0x0000ff   #   11
        self.Gate_BITS = EPD_HEIGHT
        if self.width < 128:
            self.Source_BITS = 128
        else:
            self.Source_BITS = self.width

        
    # Hardware reset
    def reset(self):
        epdconfig.digital_write(self.reset_pin, 1)
        epdconfig.delay_ms(200) 
        epdconfig.digital_write(self.reset_pin, 0)         # module reset
        epdconfig.delay_ms(2)
        epdconfig.digital_write(self.reset_pin, 1)
        epdconfig.delay_ms(200)   

    def send_command(self, command):
        epdconfig.digital_write(self.dc_pin, 0)
        epdconfig.digital_write(self.cs_pin, 0)
        epdconfig.spi_writebyte([command])
        epdconfig.digital_write(self.cs_pin, 1)

    def send_data(self, data):
        epdconfig.digital_write(self.dc_pin, 1)
        epdconfig.digital_write(self.cs_pin, 0)
        epdconfig.spi_writebyte([data])
        epdconfig.digital_write(self.cs_pin, 1)
        
    def ReadBusy(self):
        logger.debug("e-Paper busy H")
        epdconfig.delay_ms(100)
        while(epdconfig.digital_read(self.busy_pin) == 0):      # 0: idle, 1: busy
            epdconfig.delay_ms(5)
        logger.debug("e-Paper busy release")
        
    def SetWindow(self):
        self.send_command(0x61) # SET_RAM_X_ADDRESS_START_END_POSITION
        # x point must be the multiple of 8 or the last 3 bits will be ignored
        self.send_data(self.Source_BITS/256)
        self.send_data(self.Source_BITS%256)
        self.send_data(self.Gate_BITS/256)
        self.send_data(self.Gate_BITS%256)

    def TurnOnDisplay(self):
        self.send_command(0x12) # DISPLAY_REFRESH
        self.send_data(0X00)
        self.ReadBusy()
        
    def init(self):
        if (epdconfig.module_init() != 0):
            return -1
        # EPD hardware init start

        self.reset()
        
        self.ReadBusy()
        self.send_command(0x4D)
        self.send_data(0x78)

        self.send_command(0x00)
        self.send_data(0x0F)
        self.send_data(0x29)

        self.send_command(0x01)
        self.send_data(0x07)
        self.send_data(0x00)

        self.send_command(0x03)
        self.send_data(0x10)
        self.send_data(0x54)
        self.send_data(0x44)

        self.send_command(0x06)
        self.send_data(0x05)
        self.send_data(0x00)
        self.send_data(0x3F)
        self.send_data(0x0A)
        self.send_data(0x25)
        self.send_data(0x12)
        self.send_data(0x1A)

        self.send_command(0x50)
        self.send_data(0x37)

        self.send_command(0x60)
        self.send_data(0x02)
        self.send_data(0x02)
        
        self.SetWindow()
        
        self.send_command(0xE7)
        self.send_data(0x1C)

        self.send_command(0xE3)
        self.send_data(0x22)

        self.send_command(0xB4)
        self.send_data(0xD0)
        self.send_command(0xB5)
        self.send_data(0x03)

        self.send_command(0xE9)
        self.send_data(0x01)
        
        self.send_command(0x30)
        self.send_data(0x08)
        
        self.send_command(0x04)
        self.ReadBusy()
        return 0

    def getbuffer(self, image):
        # Create a pallette with the 4 colors supported by the panel
        pal_image = Image.new("P", (1,1))
        pal_image.putpalette( (0,0,0,  255,255,255,  255,255,0,   255,0,0) + (0,0,0)*252)

        # Check if we need to rotate the image
        imwidth, imheight = image.size
        if(imwidth == self.width and imheight == self.height):
            image_temp = image
        elif(imwidth == self.height and imheight == self.width):
            image_temp = image.rotate(90, expand=True)
        else:
            logger.warning("Invalid image dimensions: %d x %d, expected %d x %d" % (imwidth, imheight, self.width, self.height))

        # Convert the soruce image to the 4 colors, dithering if needed
        image_4color = image_temp.convert("RGB").quantize(palette=pal_image)
        buf_4color = bytearray(image_4color.tobytes('raw'))

        # into a single byte to transfer to the panel
        if self.width % 4 == 0 :
            Width = self.width // 4
        else :
            Width = self.width // 4 + 1
        Height = self.height 
        buf = [0x00] * int(Width * Height)
        idx = 0
        for j in range(0, Height):
            for i in range(0, Width):
                if i == Width -1:
                    buf[i + j * Width] = (buf_4color[idx] << 6) + (buf_4color[idx+1] << 4)
                    idx = idx + 2
                else:
                    buf[i + j * Width] = (buf_4color[idx] << 6) + (buf_4color[idx+1] << 4) + (buf_4color[idx+2] << 2) + buf_4color[idx+3]
                    idx = idx + 4
        return buf

    def display(self, image):
        if self.width % 4 == 0 :
            Width = self.width // 4
        else :
            Width = self.width // 4 + 1
        Height = self.height

        self.send_command(0x10)
        for j in range(0, Height):
            for i in range(0, self.Source_BITS//4):
                if i < 31 :
                    self.send_data(image[i + j * Width])
                else :
                    self.send_data(0x00)
                    
        self.TurnOnDisplay()
        
    def Clear(self, color=0x55):
        Width = self.Source_BITS//4
        Height = self.height


        self.send_command(0x10)
        for j in range(0, Height):
            for i in range(0, Width):
                    self.send_data(color)
        self.TurnOnDisplay()

    def sleep(self):
        self.send_command(0x02) # POWER_OFF
        self.ReadBusy()
        epdconfig.delay_ms(100)
        
        self.send_command(0x07) # DEEP_SLEEP
        self.send_data(0XA5)
        
        epdconfig.delay_ms(2000)
        epdconfig.module_exit()
### END OF FILE ###

